File: sdk/javascript/arkvoid.js
/**
 * ARKVOID JavaScript SDK
 * Works in Node.js and browsers.
 * 
 * Install: npm install arkvoid-sdk
 * Or use directly: copy this file
 */

const ARKVOID_BASE_URL = 'https://arkvoid.cherazen.com/api/v1'

class ArkvoidClient {
  constructor({ apiKey, agent = null, silent = false } = {}) {
    if (!apiKey?.startsWith('ARK_')) {
      throw new Error('API key must start with ARK_. Get yours at arkvoid.cherazen.com')
    }
    this.apiKey = apiKey
    this.defaultAgent = agent
    this.silent = silent
  }
  
  async trace({
    action,
    riskLevel = 'low',
    agent = null,
    riskScore = null,
    metadata = null,
    durationMs = null,
  }) {
    const agentSlug = agent || this.defaultAgent
    if (!agentSlug) {
      if (this.silent) return null
      throw new Error('Agent slug required.')
    }
    
    const payload = {
      agent_slug: agentSlug,
      action,
      risk_level: riskLevel,
      ...(riskScore !== null && { risk_score: Math.max(0, Math.min(100, riskScore)) }),
      ...(metadata && { metadata }),
      ...(durationMs !== null && { duration_ms: durationMs }),
    }
    
    try {
      const res = await fetch(`${ARKVOID_BASE_URL}/traces`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
          'User-Agent': 'arkvoid-js/1.0.0',
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(10000),
      })
      
      if (!res.ok) {
        const err = await res.json().catch(() => ({}))
        throw new Error(err.error || `HTTP ${res.status}`)
      }
      
      return await res.json()
    } catch (e) {
      if (!this.silent) {
        console.warn(`[ARKVOID] Failed to send trace:`, e.message)
      }
      return null
    }
  }
  
  // Wrap any async function with automatic tracing
  wrap(fn, options = {}) {
    const { action = fn.name, riskLevel = 'low' } = options
    const client = this
    return async function(...args) {
      const start = Date.now()
      try {
        const result = await fn.apply(this, args)
        client.trace({
          action: action || 'function_call',
          riskLevel,
          durationMs: Date.now() - start,
          metadata: { success: true, function: fn.name }
        })
        return result
      } catch (error) {
        client.trace({
          action: action || 'function_call',
          riskLevel: 'high',
          durationMs: Date.now() - start,
          metadata: { success: false, error: error.message, function: fn.name }
        })
        throw error
      }
    }
  }
}

// Simple one-liner helper
function createArkvoidTracer(apiKey, defaultAgent) {
  const client = new ArkvoidClient({ apiKey, agent: defaultAgent, silent: true })
  return (action, options = {}) => client.trace({ action, ...options })
}

module.exports = { ArkvoidClient, createArkvoidTracer }

File: sdk/python/arkvoid.py
"""
ARKVOID Python SDK
Monitor and govern your AI agents with cryptographic audit trails.
Domain: arkvoid.cherazen.com

Install: pip install requests
Usage: from arkvoid import ArkvoidClient, trace
"""

import hashlib
import json
import time
import functools
from typing import Optional, Dict, Any, Literal
from datetime import datetime, timezone

try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False
    import urllib.request
    import urllib.error

RiskLevel = Literal["low", "medium", "high", "critical"]

class ArkvoidClient:
    """
    ARKVOID governance client.
    
    Usage:
        client = ArkvoidClient(api_key="ARK_your_key_here")
        client.trace(
            agent="my-agent",
            action="document_analysis",
            risk_level="low"
        )
    """
    
    BASE_URL = "https://arkvoid.cherazen.com/api/v1"
    
    def __init__(self, api_key: str, agent: Optional[str] = None, 
                 silent: bool = False):
        """
        Args:
            api_key: Your ARKVOID API key (starts with ARK_)
            agent: Default agent slug (can override per trace)
            silent: If True, suppress all errors (fire-and-forget mode)
        """
        if not api_key.startswith("ARK_"):
            raise ValueError("API key must start with ARK_. Get yours at arkvoid.cherazen.com")
        
        self.api_key = api_key
        self.default_agent = agent
        self.silent = silent
        self._session = requests.Session() if HAS_REQUESTS else None
        if self._session:
            self._session.headers.update({
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "User-Agent": "arkvoid-python/1.0.0"
            })
    
    def trace(
        self,
        action: str,
        risk_level: RiskLevel = "low",
        agent: Optional[str] = None,
        risk_score: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None,
        input_data: Optional[Any] = None,  # will be hashed
        output_data: Optional[Any] = None,  # will be hashed
        duration_ms: Optional[int] = None,
    ) -> Optional[Dict]:
        """
        Send a trace to ARKVOID.
        
        Args:
            action: What your agent did (e.g., "document_analysis", "data_access")
            risk_level: "low" | "medium" | "high" | "critical"
            agent: Agent slug (uses default if not provided)
            risk_score: Numeric 0-100 (optional, auto-calculated if not provided)
            metadata: Any extra data to attach (model, tokens, etc.)
            input_data: Input to your function (will be SHA-256 hashed, not stored raw)
            output_data: Output of your function (will be SHA-256 hashed)
            duration_ms: How long the action took
            
        Returns:
            {"trace_id": "ark_xxx", "status": "verified", "hash": "sha256:..."} or None
        """
        agent_slug = agent or self.default_agent
        if not agent_slug:
            if self.silent:
                return None
            raise ValueError("Agent slug required. Pass agent= or set default in ArkvoidClient(agent=...)")
        
        payload = {
            "agent_slug": agent_slug,
            "action": action,
            "risk_level": risk_level,
        }
        
        if risk_score is not None:
            payload["risk_score"] = max(0, min(100, risk_score))
        
        if metadata:
            payload["metadata"] = metadata
            
        if duration_ms is not None:
            payload["duration_ms"] = duration_ms
            
        if input_data is not None:
            payload["input_hash"] = hashlib.sha256(
                json.dumps(input_data, sort_keys=True, default=str).encode()
            ).hexdigest()
            
        if output_data is not None:
            payload["output_hash"] = hashlib.sha256(
                json.dumps(output_data, sort_keys=True, default=str).encode()
            ).hexdigest()
        
        return self._send(payload)
    
    def _send(self, payload: Dict) -> Optional[Dict]:
        try:
            if HAS_REQUESTS:
                response = self._session.post(
                    f"{self.BASE_URL}/traces",
                    json=payload,
                    timeout=10
                )
                response.raise_for_status()
                return response.json()
            else:
                # Fallback using urllib
                data = json.dumps(payload).encode()
                req = urllib.request.Request(
                    f"{self.BASE_URL}/traces",
                    data=data,
                    headers={
                        "Authorization": f"Bearer {self.api_key}",
                        "Content-Type": "application/json",
                    }
                )
                with urllib.request.urlopen(req, timeout=10) as resp:
                    return json.loads(resp.read())
        except Exception as e:
            if not self.silent:
                print(f"[ARKVOID] Warning: Failed to send trace: {e}")
            return None


def trace(
    agent: str,
    action: Optional[str] = None,
    risk_level: RiskLevel = "low",
    api_key: Optional[str] = None,
    silent: bool = True,
):
    """
    Decorator for automatic function tracing.
    
    Usage:
        import os
        from arkvoid import trace
        
        @trace(agent="my-agent", api_key=os.environ["ARKVOID_API_KEY"])
        def analyze_document(doc_text):
            # Your existing code — unchanged
            return analysis
    """
    import os
    key = api_key or os.environ.get("ARKVOID_API_KEY")
    if not key and not silent:
        raise ValueError("Set ARKVOID_API_KEY env variable or pass api_key=")
    
    client = ArkvoidClient(api_key=key or "ARK_missing", agent=agent, silent=silent)
    
    def decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            start = time.time()
            result = None
            error = None
            try:
                result = func(*args, **kwargs)
                return result
            except Exception as e:
                error = e
                raise
            finally:
                duration = int((time.time() - start) * 1000)
                func_action = action or func.__name__.replace("_", " ")
                client.trace(
                    action=func_action,
                    risk_level=risk_level,
                    duration_ms=duration,
                    metadata={
                        "function": func.__name__,
                        "error": str(error) if error else None,
                        "success": error is None,
                    }
                )
        return wrapper
    return decorator
